File: 2.两数相加.java
/*
 * @Author: your name
 * @Date: 2020-04-02 06:35:51
 * @LastEditTime: 2020-04-03 06:32:11
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: /leetcode/2.两数相加.java
 */
/*
 * @lc app=leetcode.cn id=2 lang=java
 *
 * [2] 两数相加
 *
 * https://leetcode-cn.com/problems/add-two-numbers/description/
 *
 * algorithms
 * Medium (33.89%)
 * Total Accepted:    121.2K
 * Total Submissions: 357.5K
 * Testcase Example:  '[2,4,3]\n[5,6,4]'
 *
 * 给出两个 非空 的链表用来表示两个非负的整数。其中，它们各自的位数是按照 逆序 的方式存储的，并且它们的每个节点只能存储 一位 数字。
 * 
 * 如果，我们将这两个数相加起来，则会返回一个新的链表来表示它们的和。
 * 
 * 您可以假设除了数字 0 之外，这两个数都不会以 0 开头。
 * 
 * 示例：
 * 
 * 输入：(2 -> 4 -> 3) + (5 -> 6 -> 4)
 * 输出：7 -> 0 -> 8
 * 原因：342 + 465 = 807
 * 
 * 
 */
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */
// 这道题目并不是什么难题，算法很简单，链表这种数据结构也不难
// 就是创建一个新的链表，然后把输入的两个链表从头往后撸，每两个相加
// 添加一个新节点到链表后面，为了避免两个输入链表为空我们需要创建一个dummmy节点
// 将两个节点相加的新节点按照顺序加到dummmy之后，由于dummy节点本身不能变化
// 所以用一个指针cur来指向新链表的最后一个节点。
// 这道题好在最低位在链表的开头，所以可以在遍历链表的同时按从低到高的顺序直接相加
//
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1);
        ListNode cur = dummy;
        int carry = 0;
        while (l1 != null || l2 != null) {
            int d1 = l1 == null ? 0 : l1.val;
            int d2 = l2 == null ? 0 : l2.val;
            int sum = d1 + d2 + carry;
            carry = sum >= 10 ? 1 : 0;
            cur.next = new ListNode(sum % 10);
            cur = cur.next;
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }
        if (carry == 1) cur.next = new ListNode(1);
        return dummy.next;
    }
}

File: 2.两数相加.js
/*
 * @lc app=leetcode.cn id=2 lang=javascript
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
// 这道题目给定两个链表，让我们从头到尾依次往后加
// 主要注意的是进制相关的内容
var addTwoNumbers = function (l1, l2) {
  // 创建一个虚拟头结点来作为新的创建链表的头结点的前置节点
  let dummy = new ListNode(-1);
  // 这个节点是不能动的，我们把它赋值给一个变量cur
  let cur = dummy;
  // 还需要一个变量用于计算sum 这个值 val1+val2+carry
  let sum = 0;
  // 进位
  let carry = 0;

  // 循环的条件是 l1 和 l2 只要有一个不为空就可以
  // 在循环的内部需要注意判空
  while (l1 || l2) {
    sum = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + carry;
    // cur 是作为移动的那个目标前进的。
    cur.next = new ListNode(sum % 10);
    cur = cur.next;
    // 这个进制 如果是 大于等于10 则进制是1 否则是0
    carry = sum >= 10 ? 1 : 0
    // 然后开始移动指针 
    if (l1 !== null) {
      l1 = l1.next;
    }
    if (l2 !== null) {
      l2 = l2.next;
    }
  }
  // 循环结束 add 如果为1 在后面加一个add
  carry && (cur.next = new ListNode(carry));
  // 返回 头结点
  return dummy.next;
};
// @lc code=end

